code/modules/research/ordnance/doppler_array.dm

```diff
@@ -0,0 +1,315 @@
+	if(panel_open)
+		return FALSE
+	return ..()
+
+/obj/machinery/doppler_array/update_overlays()
+	. = ..()
+	if(panel_open)
+		. += mutable_appearance(icon, "[base_icon_state]_open")
+	else
+		. += mutable_appearance(icon,"[base_icon_state]_cable")
+
+	if(machine_stat & BROKEN) // Probably meant to be used on an indestructible doppler, but this is not implemented.
+		. += mutable_appearance(icon, "[base_icon_state]_screen-broken")
+	else if (machine_stat & NOPOWER)
+		. += mutable_appearance(icon, "[base_icon_state]_screen-off")
+
+/obj/machinery/doppler_array/on_deconstruction()
+	eject_disk()
+	. = ..()
+
+/obj/machinery/doppler_array/Destroy()
+	inserted_disk = null
+	QDEL_NULL(records) //We only want the list nuked, not the contents.
+	. = ..()
+
+/obj/machinery/doppler_array/proc/update_doppler_light()
+	SIGNAL_HANDLER
+	set_light_on(!(machine_stat & NOPOWER))
+
+/obj/machinery/doppler_array/AltClick(mob/user)
+	return ..() // This hotkey is BLACKLISTED since it's used by /datum/component/simple_rotation
+
+/obj/machinery/doppler_array/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "DopplerArray", name)
+		ui.open()
+
+/obj/machinery/doppler_array/ui_data(mob/user)
+	var/list/data = list()
+	data["records"] = list()
+	data["disk"] = inserted_disk?.name
+	data["storage"] = "[inserted_disk?.used_capacity] / [inserted_disk?.max_capacity] GQ"
+	for(var/datum/data/tachyon_record/singular_record in records)
+		var/list/record_data = list(
+			"name" = singular_record.name,
+			"timestamp" = singular_record.timestamp,
+			"coordinates" = singular_record.coordinates,
+			"displacement" = singular_record.displacement,
+			"factual_epicenter_radius" = singular_record.factual_radius["epicenter_radius"],
+			"factual_outer_radius" = singular_record.factual_radius["outer_radius"],
+			"factual_shockwave_radius" = singular_record.factual_radius["shockwave_radius"],
+			"theory_epicenter_radius" = singular_record.theory_radius["epicenter_radius"],
+			"theory_outer_radius" = singular_record.theory_radius["outer_radius"],
+			"theory_shockwave_radius" = singular_record.theory_radius["shockwave_radius"],
+			"reaction_results" = list(),
+			"ref" = REF(singular_record)
+		)
+
+		var/list/reaction_data = singular_record.reaction_results
+		// Make sure the list is indexed first.
+		if(reaction_data.len)
+			for (var/path in reaction_data[TANK_RESULTS_REACTION])
+				var/datum/gas_reaction/reaction_path = path
+				record_data["reaction_results"] += initial(reaction_path.name)
+			if(TANK_MERGE_OVERPRESSURE in reaction_data[TANK_RESULTS_MISC])
+				record_data["reaction_results"] += "Tank overpressurized before reaction"
+
+		data["records"] += list(record_data)
+	return data
+
+/obj/machinery/doppler_array/ui_act(action, list/params)
+	. = ..()
+	if(.)
+		return
+
+	switch(action)
+		if("delete_record")
+			var/datum/data/tachyon_record/record = locate(params["ref"]) in records
```

Comment:
should this actually like, delete the record? or do they free in the same way as gasmixes

Reply:
Initially, I'm p sure it's freed like gasmixes. But with my PR this explicitly needs to be not deleted, since disks contain references to the reaction record.